### PR TITLE
Grpc stats

### PR DESCRIPTION
**Pull request by bouda1**
# Pull Request Template

## Description

Here we have new functionalities:
* A function GetStats not finished but working
* A function ProcessCheckServiceResult working better than the external command

They are accessible from the gRPC exposed by engine. There is an example to test them in enginerpc/examples/

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 18.10.x
- [ ] 19.04.x
- [ ] 19.10.x
- [X] 20.04.x (master)

### CODE REVIEW

File: enginerpc/engine.proto [hunk 1]
```diff
@@ -1,13 +1,52 @@
...
 
 import "google/protobuf/empty.proto";
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
 
 service Engine {
   rpc GetVersion(google.protobuf.Empty) returns (Version) {}
+  rpc GetStats(google.protobuf.Empty) returns (Stats) {}
+
+  rpc ProcessServiceCheckResult(ServiceCheck) returns (CommandSuccess) {}
```
Comment (SylvestreG) [inline enginerpc/engine.proto]: Perhaps you can add a comment  to update this method further using this :

ProcessCheckResult(Check)

using this message

message Check {
  oneof check {
    HostCheck host =1;
    ServiceCheck svc =2;
  }
}

File: inc/com/centreon/engine/events/defines.hh [hunk 2]
```diff
@@ -42,6 +42,7 @@
 #define EVENT_EXPIRE_COMMENT 15  // removes expired comments
 #define EVENT_EXPIRE_HOST_ACK 16     // remove expired host acknowledgement
 #define EVENT_EXPIRE_SERVICE_ACK 17  // remove expired service acknowledgement
+#define EVENT_ENGINERPC_CHECK 18 // EngineRPC command check
```
Comment (SylvestreG) [inline inc/com/centreon/engine/events/defines.hh]: Just to be sure this define is needed by timed_event to handle grpc command ?

File: inc/com/centreon/engine/nebcallbacks.hh [hunk 3]
```diff
@@ -68,7 +68,9 @@
 #define NEBCALLBACK_ADAPTIVE_ESCALATION_DATA 40
 #define NEBCALLBACK_ADAPTIVE_TIMEPERIOD_DATA 41
 
-#define NEBCALLBACK_NUMITEMS 42 /* Total number of callback types we have. */
+#define NEBCALLBACK_ENGINERPC 42
```
Comment (SylvestreG) [inline inc/com/centreon/engine/nebcallbacks.hh]: is this needed ?

File: src/utils.cc [hunk 4]
```diff
@@ -436,28 +436,18 @@ void free_memory(nagios_macros* mac) {
   // Free memory allocated to downtimes.
   downtimes::downtime_manager::instance().clear_scheduled_downtimes();
 
-  // Free memory for the high priority event list.
-  for (timed_event_list::iterator it(timed_event::event_list_high.begin()),
-       end(timed_event::event_list_high.end());
-       it != end;) {
```
Comment (SylvestreG) [inline src/utils.cc]: I think this is leftover from another commit

File: src/xsddefault.cc [hunk 5]
```diff
@@ -109,6 +109,7 @@ int xsddefault_save_status_data() {
 
   // get number of items in the command buffer
   if (config->check_external_commands()) {
+    //FIXME DBR
```
Comment (SylvestreG) [inline src/xsddefault.cc]: xsd file are still used ?

File: inc/com/centreon/engine/events/defines.hh (hunk 2 above)
Comment (bouda1) [reply inc/com/centreon/engine/events/defines.hh]: It is the way used to force to empty stacks when we have some time. So yes, this is really needed.

File: inc/com/centreon/engine/nebcallbacks.hh (hunk 3 above)
Comment (bouda1) [reply inc/com/centreon/engine/nebcallbacks.hh]: I would like to say no, but we have a generic code that is not really ... generic... :smiley:

File: src/utils.cc (hunk 4 above)
Comment (bouda1) [reply src/utils.cc]: This just because the branch was build on top of the one with segfault already merged :smile:

File: src/xsddefault.cc (hunk 5 above)
Comment (bouda1) [reply src/xsddefault.cc]: I think so, not totally sure

File: src/utils.cc (hunk 4 above)
Comment (bouda1) [reply src/utils.cc]: The rebase solved this

**Pull request merged**